src/session_server_ssh.c

```diff
@@ -907,33 +948,248 @@ nc_sshcb_auth_password(struct nc_session *session, ssh_message msg)
+        t = msg[i]->msg_style;
+        if ((t != PAM_PROMPT_ECHO_OFF) && (t != PAM_PROMPT_ECHO_ON) && (t != PAM_TEXT_INFO) && (t != PAM_ERROR_MSG)) {
+            ERR(NULL, "PAM conversation callback received an unexpected type of message.");
+            r = PAM_CONV_ERR;
+            goto cleanup;
+        }
+    }
+
+    /* display messages with errors and/or some information and count the amount of actual authentication challenges */
+    for (i = 0; i < n_messages; i++) {
+        if (msg[i]->msg_style == PAM_TEXT_INFO) {
+            VRB(NULL, "PAM conversation callback received a message with some information for the client (%s).", msg[i]->msg);
+            n_requests--;
+        }
+        if (msg[i]->msg_style == PAM_ERROR_MSG) {
+            ERR(NULL, "PAM conversation callback received an error message (%s).", msg[i]->msg);
+            r = PAM_CONV_ERR;
+            goto cleanup;
+        }
+    }
+
+    /* there are no requests left for the user, only messages with some information for the client were sent */
+    if (n_requests <= 0) {
+        r = PAM_SUCCESS;
+        goto cleanup;
+    }
+
+    /* it is the PAM module's responsibility to release both, this array and the responses themselves */
+    *resp = calloc(n_requests, sizeof **resp);
+    prompts = calloc(n_requests, sizeof *prompts);
+    echo = calloc(n_requests, sizeof *echo);
+    if (!(*resp) || !prompts || !echo) {
+        ERRMEM;
+        r = PAM_BUF_ERR;
+        goto cleanup;
+    }
+
+    /* set the prompts for the user */
+    j = 0;
+    for (i = 0; i < n_messages; i++) {
+        if ((msg[i]->msg_style == PAM_PROMPT_ECHO_ON) || (msg[i]->msg_style == PAM_PROMPT_ECHO_OFF)) {
+            prompts[j++] = msg[i]->msg;
+        }
+    }
+
+    /* iterate over all the messages and adjust the echo array accordingly */
+    j = 0;
+    for (i = 0; i < n_messages; i++) {
+        if (msg[i]->msg_style == PAM_PROMPT_ECHO_ON) {
+            echo[j++] = 1;
+        }
+        if (msg[i]->msg_style == PAM_PROMPT_ECHO_OFF) {
+            /* no need to set to 0 because of calloc */
+            j++;
+        }
+    }
+
+    /* print all the keyboard-interactive challenges to the user */
+    r = ssh_message_auth_interactive_request(clb_data->msg, name, instruction, n_requests, prompts, echo);
+    if (r != SSH_OK) {
+        ERR(NULL, "Failed to send an authentication request.");
+        r = PAM_CONV_ERR;
+        goto cleanup;
+    }
+
+    if (opts->auth_timeout) {
+        nc_gettimespec_mono_add(&ts_timeout, opts->auth_timeout * 1000);
+    }
+
+    /* get user's replies */
+    do {
+        if (!nc_session_is_connected(clb_data->session)) {
+            ERR(NULL, "Communication SSH socket unexpectedly closed.");
+            break;
+        }
+
+        reply = ssh_message_get(libssh_session);
+        if (reply) {
+            break;
+        }
+
+        usleep(NC_TIMEOUT_STEP);
+    } while ((opts->auth_timeout) && (nc_difftimespec_cur(&ts_timeout) >= 1));
```

Comment:
There seem to be no error message on timeout while there is one for other conditions. My suggestion would be to treat error on line 1060 explicitly and treat condition on 1072 as timeout only.